### DISCUSSION TITLE
[C.Deliv] Environment allocation

### DISCUSSION THREAD

**Discussion by erkist** [cdfoundation/sig-events#38] [General]
In PR https://github.com/cdfoundation/sig-events/pull/23 the concept of Environment Allocation, with a corresponding event, is brought up and defined as follows:

> Announces that the allocation status of an environment has changed (e.g. it has been allocated, de-allocated or taken offline).
> 
> Environment Allocation Status Changed Events MUST include the following attributes:
> 
> - **Environment ID**: Unique identifier for the environment.
> - **Allocation Status**: The new allocation status for the environment (Allocated, Idle, Offline)
>
> Optional attributes:
> 
> - **Environment Name**: Machine-friendly and human-friendly name of recipe, e.g. "gemini-backbone-hil-4".
> - **Environment Display Name**: UX/human-friendly display name of recipe, e.g. "Gemini Backbone HIL 4”.
> - **Environment URL**: URL to reference where the environment is located.

And further clarified by me as:

> To explain a bit more what I am after:
> Step 1: Every activity or test that needs to run may declare what it needs from the environment (ESCI calls this a Resource Set Capability, Eiffel calls this a Constraint )
>
> Step 2: When this activity or test needs to be run, we need to find a suitable environment to run this activity/test on. This is referred to above as Environment Allocation.
>
> Step 3 option 1: If an environment with the right capabilities exists and is currently idle, an event is immediately sent out by the environment allocation service stating "This environment is now allocated to this activity", and an orchestration server can start running.
>
> Step 3 option 2: If an environment with the right capabilities exists but is currently allocated to something else, the orchestration server just waits for the event from 3a (note that other activities/tests may wait for the same environment and get to run first)
>
> Step 3 option 3: If no environment with the right capabilities exist, the actor trying to put the request in the environment allocation queue is immediately informed that the request cannot be fulfilled, and the pipeline is aborted.

Input to the discussion:
* [Resource Set (specifially the paragraph about Capabilities) in ESCI](https://github.com/erkist/sig-interoperability/tree/erkist-esci/workstreams/events_in_cicd/input/enterprise-scale-ci#resource-set)

Do we want to represent this concept in the spec? If so, how? What events do we want to have attached to it?

**Comment by m-linner-ericsson** [comment]
Even though I think this is an advanced concept there are several situations where this could happen.

* Hardware - some tests might need to run a specific hardware
* Optimization - some test will do better if they run on a specific hardware e.g. a docker image might run on any node but will run significantly faster if run on a node with a GPU
* Limits - while developing a tool you might use a small cluster but when running the performance and stress test you need to use a large cluster

For the wording I guess [allocate](https://en.wiktionary.org/wiki/allocate) and [deallocate](https://en.wiktionary.org/wiki/deallocate) should work.
